src/Scrutor/AttributeSelector.cs

```diff
@@ -22,13 +23,55 @@ public void Populate(IServiceCollection services)
 
                 foreach (var attribute in typeInfo.GetCustomAttributes<ServiceDescriptorAttribute>())
                 {
-                    var serviceType = attribute.ServiceType ?? type;
+                    var serviceType = GetServiceType(typeInfo, attribute);
 
                     var descriptor = new ServiceDescriptor(serviceType, type, attribute.Lifetime);
 
                     services.Add(descriptor);
                 }
             }
         }
+
+        private Type GetServiceType(TypeInfo typeInfo, ServiceDescriptorAttribute attribute)
+        {
+            var serviceType = attribute.ServiceType;
+            if (ReferenceEquals(null, serviceType))
+            {
+                return typeInfo.AsType();
+            }
+
+            if (!ValidateInheritence(typeInfo, serviceType))
```

Comment:
This logic seems awfully complex compared to just doing `serviceType.GetTypeInfo().IsAssignableFrom(typeInfo)`, no? That would also take care of co- and contravariance. Have I missed something?